File: sdk/desktop/windows/tests/Kseal.Desktop.Tests/EnterprisePolicyWiringTests.cs
using Kseal.Desktop.Pe;
using Xunit;

namespace Kseal.Desktop.Tests;

/// <summary>
/// Verifies the effect of enterprise controls on the live client against the
/// real Rust core with a faked OS environment.
/// </summary>
public class EnterprisePolicyWiringTests
{
    private sealed class FixedClock(long millis) : IClock
    {
        public long NowMillis() => millis;
    }

    private static (KsealDesktopClient Client, FakeWindowsEnvironment Env, BufferingTelemetrySink Sink) NewClient(
        EnterprisePolicy enterprise,
        bool proofKeyIsHardwareBacked = false,
        IReadOnlySet<string>? enabledProbes = null)
    {
        var env = new FakeWindowsEnvironment
        {
            Authenticode = new AuthenticodeInfo(true, true, "CN=Contoso", "ABCD", true),
            Pe = PeImage.Parse(PeBuilder.BuildText([1, 2, 3, 4])),
        };
        var core = NativeTrustCore.Create(new byte[32], new byte[32]);
        var sink = new BufferingTelemetrySink();
        var opts = new KsealDesktopOptions { BuildHash = "build-hash", EnabledProbes = enabledProbes };
        var client = new KsealDesktopClient(
            core, env, opts, new FileConfigProvider(Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString())),
            sink, new LocalCodeIntegrityAttestor(), "install-hash", new FixedClock(0),
            proofKeyIsHardwareBacked, enterprise);
        return (client, env, sink);
    }

    [Fact]
    public void StrictPolicyFlagsForeignModule()
    {
        var (client, env, _) = NewClient(EnterprisePolicy.Strict);
        env.Foreign.Add(@"C:\temp\inject.dll");
        Assert.Contains(RiskSignal.Hooking, client.EvaluateRisk().Signals);
    }

    [Fact]
    public void InjectionAllowlistSuppressesSanctionedModule()
    {
        var policy = new EnterprisePolicy { InjectionAllowlist = [@"C:\Program Files\Acme\"] };
        var (client, env, _) = NewClient(policy);
        env.Foreign.Add(@"C:\Program Files\Acme\plugin.dll");
        Assert.DoesNotContain(RiskSignal.Hooking, client.EvaluateRisk().Signals);
    }

    [Fact]
    public void AllowlistStillFlagsNonListedModule()
    {
        var policy = new EnterprisePolicy { InjectionAllowlist = [@"C:\Program Files\Acme\"] };
        var (client, env, _) = NewClient(policy);
        env.Foreign.Add(@"C:\Program Files\Acme\plugin.dll");
        env.Foreign.Add(@"C:\temp\evil.dll");
        Assert.Contains(RiskSignal.Hooking, client.EvaluateRisk().Signals);
    }

    [Fact]
    public void PermitDebuggerDropsDebuggerProbeEvenWhenEnabled()
    {
        var enabled = new HashSet<string> { "windows.debugger" };

        var (strict, env1, _) = NewClient(EnterprisePolicy.Strict, enabledProbes: enabled);
        env1.Debugger = true;
        Assert.Contains(RiskSignal.Debugger, strict.EvaluateRisk().Signals);

        var (permitted, env2, _) = NewClient(
            new EnterprisePolicy { PermitDebugger = true }, enabledProbes: enabled);
        env2.Debugger = true;
        Assert.DoesNotContain(RiskSignal.Debugger, permitted.EvaluateRisk().Signals);
    }

    [Fact]
    public void RequireHardwareBackedProofKeyRaisesSecureHwMissing()
    {
        var policy = new EnterprisePolicy { RequireHardwareBackedProofKey = true };

        var (missing, _, _) = NewClient(policy, proofKeyIsHardwareBacked: false);
        Assert.Contains(RiskSignal.SecureHwMissing, missing.EvaluateRisk().Signals);

        var (present, _, _) = NewClient(policy, proofKeyIsHardwareBacked: true);
        Assert.DoesNotContain(RiskSignal.SecureHwMissing, present.EvaluateRisk().Signals);

        var (off, _, _) = NewClient(EnterprisePolicy.Strict, proofKeyIsHardwareBacked: false);
        Assert.DoesNotContain(RiskSignal.SecureHwMissing, off.EvaluateRisk().Signals);
    }

    [Fact]
    public void MinimalVerbosityDropsCleanEvents()
    {
        var (client, _, sink) = NewClient(new EnterprisePolicy { TelemetryVerbosity = TelemetryVerbosity.Minimal });
        client.ReportEvent(EventType.PolicyDecision);
        client.FlushTelemetry();
        Assert.Empty(sink.Drain());
    }

    [Fact]
    public void StandardVerbosityKeepsCleanEvents()
    {
        var (client, _, sink) = NewClient(EnterprisePolicy.Strict);
        client.ReportEvent(EventType.PolicyDecision);
        client.FlushTelemetry();
        Assert.NotEmpty(sink.Drain());
    }
}

File: sdk/desktop/macos/Tests/KsealDesktopTests/EnterprisePolicyWiringTests.swift
import XCTest
@testable import KsealDesktop

private struct FixedClock: Clock {
    let millis: Int64
    func nowMillis() -> Int64 { millis }
}

/// Verifies the effect of enterprise controls on the live SDK against the real
/// Rust core with a controlled environment.
final class EnterprisePolicyWiringTests: XCTestCase {

    private func makeSDK(
        env: DesktopEnvironment,
        enterprise: EnterprisePolicy,
        sink: BufferingTelemetrySink = BufferingTelemetrySink(),
        enabledProbes: Set<String>? = nil,
        proofKeyIsHardwareBacked: Bool = false
    ) throws -> KsealDesktop {
        let policy = DesktopIntegrityPolicy(
            expectedTeamIdentifier: "ABCDE12345",
            expectedSigningIdentifier: "com.example.app"
        )
        let core = try NativeTrustCore.create(
            configPublicKey: Data(repeating: 7, count: 32),
            proofKey: Data("instance-key".utf8),
            platform: .desktopMac
        )
        return KsealDesktop(
            core: core,
            env: env,
            options: KsealDesktopOptions(integrityPolicy: policy, enabledProbes: enabledProbes),
            configProvider: FileConfigProvider(directory: FileManager.default.temporaryDirectory),
            telemetrySink: sink,
            attestor: LocalCodeIntegrityAttestor(),
            installIdentityHash: "deadbeef",
            clock: FixedClock(millis: 1_700_000_000_000),
            proofKeyIsHardwareBacked: proofKeyIsHardwareBacked,
            enterprisePolicy: enterprise
        )
    }

    func testStrictPolicyMatchesPreExistingBehavior() throws {
        let env = FakeDesktopEnvironment()
        env.foreignImages = ["/tmp/inject.dylib"]
        let assessment = try makeSDK(env: env, enterprise: .strict).evaluateRisk()
        XCTAssertTrue(assessment.signals.contains(.hooking))
    }

    func testInjectionAllowlistSuppressesSanctionedModule() throws {
        let env = FakeDesktopEnvironment()
        env.foreignImages = ["/Library/Acme/plugin.dylib"]
        let policy = EnterprisePolicy(injectionAllowlist: ["/Library/Acme/"])
        let assessment = try makeSDK(env: env, enterprise: policy).evaluateRisk()
        XCTAssertFalse(assessment.signals.contains(.hooking))
    }

    func testAllowlistStillFlagsNonListedModule() throws {
        let env = FakeDesktopEnvironment()
        env.foreignImages = ["/Library/Acme/plugin.dylib", "/tmp/evil.dylib"]
        let policy = EnterprisePolicy(injectionAllowlist: ["/Library/Acme/"])
        let assessment = try makeSDK(env: env, enterprise: policy).evaluateRisk()
        XCTAssertTrue(assessment.signals.contains(.hooking))
    }

    func testStrictPolicyConsultsNativeHook() throws {
        let env = FakeDesktopEnvironment()
        env.nativeHook = 1 // no allowlist configured → native scan is consulted
        let assessment = try makeSDK(env: env, enterprise: .strict).evaluateRisk()
        XCTAssertTrue(assessment.signals.contains(.hooking))
    }

    func testInjectionAllowlistSuppressesNativeHook() throws {
        let env = FakeDesktopEnvironment()
        env.nativeHook = 1 // allowlist-unaware native scan would fire ...
        let policy = EnterprisePolicy(injectionAllowlist: ["/Library/Acme/"])
        let assessment = try makeSDK(env: env, enterprise: policy).evaluateRisk()
        // ... but a configured allowlist defers to the allowlist-aware managed
        // checks, so the native scan is not consulted.
        XCTAssertFalse(assessment.signals.contains(.hooking))
    }

    func testPermitDebuggerDropsDebuggerProbeEvenWhenEnabled() throws {
        let env = FakeDesktopEnvironment()
        env.traced = true
        let enabled: Set<String> = ["macos.debugger"]
        // Without permit: the explicitly enabled debugger probe fires.
        let strict = try makeSDK(env: env, enterprise: .strict, enabledProbes: enabled).evaluateRisk()
        XCTAssertTrue(strict.signals.contains(.debugger))
        // With permit: it is suppressed for the managed developer machine.
        let permitted = try makeSDK(
            env: env, enterprise: EnterprisePolicy(permitDebugger: true), enabledProbes: enabled
        ).evaluateRisk()
        XCTAssertFalse(permitted.signals.contains(.debugger))
    }

    func testRequireHardwareBackedProofKeyRaisesSecureHwMissing() throws {
        let env = FakeDesktopEnvironment()
        let policy = EnterprisePolicy(requireHardwareBackedProofKey: true)
        // Hardware unavailable → signal raised.
        let missing = try makeSDK(env: env, enterprise: policy, proofKeyIsHardwareBacked: false).evaluateRisk()
        XCTAssertTrue(missing.signals.contains(.secureHwMissing))
        // Hardware present → not raised.
        let present = try makeSDK(env: env, enterprise: policy, proofKeyIsHardwareBacked: true).evaluateRisk()
        XCTAssertFalse(present.signals.contains(.secureHwMissing))
        // Control off → never raised even without hardware.
        let off = try makeSDK(env: env, enterprise: .strict, proofKeyIsHardwareBacked: false).evaluateRisk()
        XCTAssertFalse(off.signals.contains(.secureHwMissing))
    }

    func testMinimalVerbosityDropsCleanEvents() throws {
        let env = FakeDesktopEnvironment() // clean, signed
        let sink = BufferingTelemetrySink()
        let sdk = try makeSDK(
            env: env, enterprise: EnterprisePolicy(telemetryVerbosity: .minimal), sink: sink
        )
        sdk.reportEvent(.policyDecision)
        sdk.flushTelemetry()
        XCTAssertTrue(sink.drain().isEmpty, "clean event should be dropped at minimal verbosity")
    }

    func testStandardVerbosityKeepsCleanEvents() throws {
        let env = FakeDesktopEnvironment()
        let sink = BufferingTelemetrySink()
        let sdk = try makeSDK(env: env, enterprise: .strict, sink: sink)
        sdk.reportEvent(.policyDecision)
        sdk.flushTelemetry()
        XCTAssertFalse(sink.drain().isEmpty, "clean event should be recorded at standard verbosity")
    }
}
